File: sandbox/algm/lintcode/max-depth-binary-tree/recursion.py
"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""
class Solution:
    """
    @param root: The root of binary tree.
    @return: An integer
    """ 
    def maxDepth(self, root):
        # write your code here
        if root is None:
            return(0)
        elif root.left is None and root.right is None:
            #leaf node
            return(1)
        else:
            return(1+max(self.maxDepth(root.left),self.maxDepth(root.right)))

File: coursework/online_judge_practice/leetcode/binary tree paths/recursion.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
//time complexity: nlogn
//space: O(n) without stack space
 import java.util.*;
public class Solution {
    public List<String> binaryTreePaths(TreeNode root) {
        List<String> ret = new LinkedList<String>();
        if (root == null) return(ret);
        if (root.left == null && root.right == null) {
            ret.add(""+root.val);
        }
        
        for (String i:binaryTreePaths(root.left)) {
            ret.add(root.val+"->"+i);
        
        }
        
        for (String i:binaryTreePaths(root.right)) {
            ret.add(root.val+"->"+i);
        
        }
        return(ret);
    }
}
